rfd/0056-bot-for-cert-renewals.md

```diff
@@ -0,0 +1,702 @@
+short-lived. This leads to a few problems with renewable certificates:
+
+ 1. Renewable certificates that are not continually renewed (because the
+    job that issued them has completed, for instance) can be stolen without
+    triggering a certificate generation mismatch.
+ 2. It breaks the simple implementation of generation counters, of which each
+    user only has a single generation counter. This is solvable but not
+    worthwhile given the above: it would just allow each certificate lineage
+    to be separately stolen.
+
+Instead, we opt to have bots renew certificates by continually
+re-authenticating using the IAM join method. As EC2 joining relies on instance
+identity documents which can themselves be stolen, this join method will
+initially not be supported until a workaround is developed.
+
+[aws-join]: https://goteleport.com/docs/setup/guides/joining-nodes-aws/
+
+### Implementation
+
+The bot will be implemented as a standalone tool (ie under `tool/tbot`) and
+_not_ as a service that runs in a `teleport` process.
+
+The bulk of the new logic (parsing configuration, writing certificates to a
+destination, etc.) will be implemented in a new `renew` package, which will
+enable us to start with a standalone bot, and incorporate this functionality
+into other parts of Teleport (ie database access) in the future.
+
+#### API Objects
+
+On the backend, registering a new bot (via `tctl`) creates a set of related API
+objects:
+
+- a non-interactive User for the bot to act as (when user certs are requested)
+- a role for the bot, which will contain the minimum set of permissions
+  necessary to watch for CA rotations, as well as the ability to impersonate
+  other roles
+- a token that the bot can use to join the cluster for the first time
+
+In order to discover related objects and ensure that no "zombie" items are left behind,
+`tctl` will ensure that:
+
+- the objects related to a bot all follow a standardized naming convention that
+  includes the name of the bot
+- a new `teleport.dev/owned-by` label is applied to all resources related to a
+  particular bot
+
+Note: end users should never need to manually create, modify, or delete these
+objects. All operations on bots and their related objects is performed through
+the `tctl bots` subcommand.
+
+#### `tbot` configuration
+
+The goal is to keep tbot configuration to a minimum to make for a quick getting
+started experience, however we recognize many use cases may quickly become
+difficult to encode on the command line.
+
+To that end, we'd like to follow the philosophy of "keep simple things simple
+but make complex things possible": the simplest / most popular use cases should
+ideally require minimal (zero?) configuration beyond a concise CLI, but more
+complicated use cases can make use of a `tbot.yaml` configuration file.
+
+An example `tbot.yaml` with all default / advanced options specified:
+
+```yaml
+# Storage for the bot's private data
```

Comment:
Is is possible to set the Token & Auth server as part of tbot.yaml?

Reply:
Yes:

```yaml
auth_server: example.com:3023

onboarding:
  token: asdf1234
```

Reply:
Should we copy teleport.yaml to keep things consistent?

```
tbot:
    auth_token: asdf1234
    auth_servers:
        - "https://teleport-9a.asteroid.earth:443"
```

Reply:
Since it's a completely separate binary from the rest of Teleport we don't really need to pack everything under a `tbot` object like we do for auth / proxy / etc. Given that, it's fairly arbitrary.

Reply:
Thanks @timothyb89 . Closing this issue.